### DISCUSSION TITLE
How to Dynamically Create Workflows in Elsa Using C#?

### DISCUSSION THREAD

**Discussion by 312977** [elsa-workflows/elsa-core#4425] [Q&A]
Subject: How to Dynamically Create Workflows in Elsa Using C#?

Hello,

I'm interested in building a low-code platform using Elsa Core. However, I find that the built-in UI is too specialized for end-users, requiring them to use code to complete dynamic workflows. In my use case, the workflows are entirely user-defined and are submitted to me in a JSON format that I've specified. I then need to convert this JSON into a format that Elsa can understand.

I'm looking to dynamically create Elsa workflows using C# code, but I haven't found any documentation on how to do this. Specifically, there are no guidelines on dynamically creating workflows through code, nor are there detailed explanations for the parameters and variables involved.

Could you please guide me on how to proceed? When can we expect a demo or documentation that covers this aspect?

Thank you!

**Comment by sfmskywalker** [comment]
Hello @312977,

Thank you for bringing up this important topic.

In Elsa 3, you have the capability to dynamically provide workflows to the system by implementing the `IWorkflowProvider`. This interface allows you to read workflows from JSON and map them dynamically into Elsa workflows. This means you can convert the JSON format you receive from users into a format that Elsa understands, all through C# code.

I understand that there might not be ample guidance in this area currently. I'm glad you brought this up as it indicates there's interest in the advanced functionality that Elsa offers. To address this, I've created [an issue](https://github.com/elsa-workflows/elsa-documentation/issues/24) in the Elsa documentation repository to enhance the documentation around this topic.

For some practical examples, you might want to check out implementations like `BlobStorageWorkflowProvider` and `ClrWorkflowProvider` in the Elsa codebase. These can give you a starting point on how to implement your own workflow provider.

I hope this helps, and we'll work on improving the documentation to provide more clarity on this topic.

**Reply by Rathna314** [reply]
Hello Sipke Schoorstra,

I wish ELSA had more complex examples of how to execute a rule engine, control flows , sending json objects in the request etc..
Half of the pages show 404 and most examples don't work when we have complex business requirements. Hopefully they atleast put more documentation and samples that can be used in real time projects.

**Comment by neerajmathur** [comment]
+1 for above requirement, I am struggling with same. 
We are building a low code platform with our own basic user friendly workflow editor, JSON produced by our editor needs to be converted to ELSA format.

**Comment by 312977** [comment]
For me, examining the source code might be a bit too costly. Another approach is to utilize the existing API endpoints to achieve my desired scenario. However, I still encounter an issue: there is no corresponding documentation available for the Elsa API endpoints. At most, I can only infer from the existing UI requests. Is there any documentation available for this aspect?

Additionally, regarding the mentioned documentation for creating workflows using C# code, is there a plan for it? If so, could you provide an approximate timeline?

Furthermore, I would like to suggest that, apart from addressing critical bugs, prioritizing the enhancement of documentation should be of the highest importance. It is a low-cost measure that can help more people understand and utilize Elsa Core, thereby fostering a stronger ecosystem and potentially creating business opportunities
